### ISSUE TITLE
Atlas textures not showing up

### ISSUE THREAD

**Issue by osrf-migration** [osrf/drcsim#89]
**Original report ([archived issue](https://osrf-migration.github.io/drcsim-gh-pages/#!/osrf/drcsim/issues/89)) by Thomas Koletschka (Bitbucket: [thomasko](https://bitbucket.org/%7Bb33092ac-6376-48d1-9803-d9e6de445d1a%7D/)).**

----------------------------------------
As discussed in the comments to Nate's reply at http://answers.gazebosim.org/question/925/multiple-materials-per-link/ Atlas' textures are not being loaded in Gazebo. This can be solved by adding material scripts just like for the other textured models but while trying to do so I ran into the problem of most parts (esp. the legs) requiring multiple textures and I only managed to load one (someone with more knowledge about Ogre material scripts can probably solve this faster).

The textures show up in rviz without any additional material scripts so while both Gazebo and RViz are based on Ogre, RViz' collada importer seems to be able to handle material properties embedded into the collada files better than Gazebo's does.
For example one of the collada trees from google warehouse loads up with textures and colors without any additional material scripts while the Atlas model is completely blank (even the color on the MultiSense head is missing and needs to be added through separate material tag/script in the sdf file).

It would make the process of creating new models a lot easier if Gazebo would be able to handle all the material properties embedded into the collada file without the need to write separate material scripts. According to Nate's comment this seems to be the desired behaviour anyway and Atlas seems to be supposed to show up as textured already so it's a bug and not a feature request.

**Comment by osrf-migration** [comment]
**Original comment by Nate Koenig (Bitbucket: [Nathan Koenig](https://bitbucket.org/%7Bc862cdd9-fcc8-4419-9fc7-e20db14b8fcb%7D/)).**

----------------------------------------
Gazebo loads embedded textures as long as Gazebo can find the texture. For example, take a look at the `nist_maze_wall_120` model. The Collada file contains embedded textures, and Gazebo loads them.

I think the problem is that the Collada files for the Atlas robot are correct only for ROS, and not for Gazebo. By that I mean, we are not properly telling Gazebo where to look for textures.

You have gotten around this by telling Gazebo via material scripts in SDF. The longer term solution will be to fix the Atlas package.

Long story short: Atlas robot model is probably the culprit.

**Comment by osrf-migration** [comment]
**Original comment by Nate Koenig (Bitbucket: [Nathan Koenig](https://bitbucket.org/%7Bc862cdd9-fcc8-4419-9fc7-e20db14b8fcb%7D/)).**

----------------------------------------

* changed priority from "major" to "critical"

**Comment by osrf-migration** [comment]
**Original comment by Nate Koenig (Bitbucket: [Nathan Koenig](https://bitbucket.org/%7Bc862cdd9-fcc8-4419-9fc7-e20db14b8fcb%7D/)).**

----------------------------------------
@hsu Any ideas on this ticket?

**Comment by osrf-migration** [comment]
**Original comment by Thomas Koletschka (Bitbucket: [thomasko](https://bitbucket.org/%7Bb33092ac-6376-48d1-9803-d9e6de445d1a%7D/)).**

----------------------------------------
Textures also don't load for the new gazebo_models models like the ranger (only one i tested but I'm guessing it's the same for the other). Maybe the problem/difference is that ROS uses the .dae file's location while Gazebo uses some Gazebo dir or root?

Regarding the missing head color I guess it's due to the urdf->sdf conversion bug that dropped the material tag which already got fixed in some branch though I think (can't find the ticket right now).

I just noticed that ``nist_mase_wall_120`` uses just the filename instead of relative path to the materials/textures/ folder ... maybe that's what Gazebo is using while ros relies on a relative path?

Edit: hmmm, nope, removing the relative filepath in the .dae files and leaving just the texture name only breaks it in rviz but doesn't make it work in gazebo. Using an absolute filepath doesn't seem to work either, so the issue is something else I guess.

Edit2: what's also interesting is that Meshlab loads textures correctly for the NIST wall but not for the objects that also fail for Gazebo. So the FBX collada exporter (used for Atlas) and OpenCOLLADA exporter (used for polaris) seem to be doing something different than Blender's and SketchUp's exporters that rviz is able to handle but other programs including gazebo aren't.

**Comment by osrf-migration** [comment]
**Original comment by Thomas Koletschka (Bitbucket: [thomasko](https://bitbucket.org/%7Bb33092ac-6376-48d1-9803-d9e6de445d1a%7D/)).**

----------------------------------------
Regarding @nkoenig's question about ogre.log:

I don't get any errors related to Atlas textures, but depending on the file path i use for the texture in the NIST wall I either get an error or a success message about loading the texture. In either case it shows up in Gazebo however (even when ogre.log shows an error about not being able to load the texture):

- using ../materials/textures/plywood256x256.jpg for the collada texture tag (the file printed in the error message does exist):

        01:39:00: OGRE EXCEPTION(6:FileNotFoundException): Cannot locate resource /home/thomas/gazebo_models/nist_maze_wall_120/meshes/../materials/textures/plywood256x256.jpg in resource group General or any other group. in ResourceGroupManager::openResource at /build/buildd/ogre-1.7.4/OgreMain/src/OgreResourceGroupManager.cpp (line 753)
        01:39:00: Error loading texture /home/thomas/gazebo_models/nist_maze_wall_120/meshes/../materials/textures/plywood256x256.jpg. Texture layer will be blank. Loading the texture failed with the following exception: OGRE EXCEPTION(6:FileNotFoundException): Cannot locate resource /home/thomas/gazebo_models/nist_maze_wall_120/meshes/../materials/textures/plywood256x256.jpg in resource group General or any other group. in ResourceGroupManager::openResource at /build/buildd/ogre-1.7.4/OgreMain/src/OgreResourceGroupManager.cpp (line 753)

- using just plywood256x256.jpg for collada's texture tag:

         01:43:26: Texture: /home/thomas/gazebo_models/nist_maze_wall_120/meshes/../materials/textures/plywood256x256.jpg: Loading 1 faces(PF_R8G8B8,256x256x1) with 5 generated mipmaps from Image. Internal format is PF_X8R8G8B8,256x256x1.

**Comment by osrf-migration** [comment]
**Original comment by Ian Chen (Bitbucket: [Ian Chen](https://bitbucket.org/%7Beaa6fca5-6deb-43f6-907f-971c144735dd%7D/), GitHub: [iche033](https://github.com/iche033)).**

----------------------------------------
try this [patch](https://bitbucket.org/osrf/gazebo/pull-request/256/update-collada-texture-loader/diff) from gazebo repo

**Comment by osrf-migration** [comment]
**Original comment by Thomas Koletschka (Bitbucket: [thomasko](https://bitbucket.org/%7Bb33092ac-6376-48d1-9803-d9e6de445d1a%7D/)).**

----------------------------------------
yep, works for me, awesome :)

**Comment by osrf-migration** [comment]
**Original comment by Ian Chen (Bitbucket: [Ian Chen](https://bitbucket.org/%7Beaa6fca5-6deb-43f6-907f-971c144735dd%7D/), GitHub: [iche033](https://github.com/iche033)).**

----------------------------------------

* changed state from "new" to "resolved"

pull request 256 in osrf/gazebo repo
